File: tools/check_implementation_phases.py
"""Verify that implementation phases partition the runtime SPEC cases."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SPEC_PATH = ROOT / "docs" / "SPEC.md"
PLAN_PATH = ROOT / "config" / "implementation-phases.json"

CASE_PATTERN = re.compile(r"^- \*\*(S-[A-Z0-9-]+)\*\*", re.MULTILINE)
HEADING_CASE_PATTERN = re.compile(r"^### (S-ARM-[0-9]+)\b", re.MULTILINE)
DECLARED_LIMIT_PATTERN = re.compile(
    r"^- \*\*(S-[A-Z0-9-]+)\*\*.*Declared limit .*NOT a test case",
    re.MULTILINE,
)


def fail(message: str) -> None:
    print(f"implementation phase check failed: {message}", file=sys.stderr)
    raise SystemExit(1)


def main() -> None:
    spec = SPEC_PATH.read_text(encoding="utf-8")
    config = json.loads(PLAN_PATH.read_text(encoding="utf-8"))

    defined = CASE_PATTERN.findall(spec) + HEADING_CASE_PATTERN.findall(spec)
    duplicates = sorted({case for case in defined if defined.count(case) > 1})
    if duplicates:
        fail(f"duplicate SPEC definitions: {', '.join(duplicates)}")

    cases = set(defined)
    expected_defined = config["expectedDefinedCases"]
    if len(cases) != expected_defined:
        fail(f"expected {expected_defined} defined cases, found {len(cases)}")

    declared_limits = set(config["declaredLimits"])
    spec_declared_limits = set(DECLARED_LIMIT_PATTERN.findall(spec))
    if declared_limits != spec_declared_limits:
        fail(
            "declared-limit manifest differs from SPEC: "
            f"manifest={sorted(declared_limits)}, spec={sorted(spec_declared_limits)}"
        )
    unknown_limits = declared_limits - cases
    if unknown_limits:
        fail(f"declared limits absent from SPEC: {', '.join(sorted(unknown_limits))}")

    assignments: dict[str, list[str]] = {case: [] for case in cases}
    empty_patterns: list[str] = []
    phase_counts: dict[str, int] = {}

    for phase, raw_patterns in config["phases"].items():
        compiled = [re.compile(pattern) for pattern in raw_patterns]
        for raw_pattern, pattern in zip(raw_patterns, compiled, strict=True):
            if not any(pattern.fullmatch(case) for case in cases):
                empty_patterns.append(f"{phase}:{raw_pattern}")
        for case in cases:
            if any(pattern.fullmatch(case) for pattern in compiled):
                assignments[case].append(phase)
        phase_counts[phase] = sum(phase in owners for owners in assignments.values())

    if empty_patterns:
        fail(f"patterns matching no case: {', '.join(empty_patterns)}")

    missing = sorted(case for case, owners in assignments.items() if not owners)
    multiple = sorted(case for case, owners in assignments.items() if len(owners) > 1)
    if missing:
        fail(f"unassigned cases: {', '.join(missing)}")
    if multiple:
        detail = ", ".join(f"{case}={assignments[case]}" for case in multiple)
        fail(f"multiply assigned cases: {detail}")

    test_count = len(cases - declared_limits)
    expected_tests = config["expectedTestCases"]
    if test_count != expected_tests:
        fail(f"expected {expected_tests} test cases, found {test_count}")

    counts = " ".join(f"{phase}={count}" for phase, count in phase_counts.items())
    print(
        f"implementation phases OK: {len(cases)} definitions, "
        f"{test_count} tests, {len(declared_limits)} declared limit; {counts}"
    )


if __name__ == "__main__":
    main()

File: tools/check-implementation-phases.mjs
// Verify that implementation phases partition the runtime SPEC cases. Kept in
// Node so `npm run verify` uses the repository's pinned runtime end to end.
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const spec = readFileSync(path.join(root, "docs", "SPEC.md"), "utf8");
const config = JSON.parse(readFileSync(path.join(root, "config", "implementation-phases.json"), "utf8"));

function fail(message) {
  process.stderr.write(`implementation phase check failed: ${message}\n`);
  process.exit(1);
}

function matches(pattern, text) {
  return [...text.matchAll(pattern)].map(match => match[1]);
}

const defined = [
  ...matches(/^- \*\*(S-[A-Z0-9-]+)\*\*/gm, spec),
  ...matches(/^### (S-ARM-[0-9]+)\b/gm, spec),
];
const duplicates = [...new Set(defined.filter((value, index) => defined.indexOf(value) !== index))].sort();
if (duplicates.length > 0) fail(`duplicate SPEC definitions: ${duplicates.join(", ")}`);

const cases = new Set(defined);
if (cases.size !== config.expectedDefinedCases) fail(`expected ${String(config.expectedDefinedCases)} defined cases, found ${String(cases.size)}`);

const declaredLimits = new Set(config.declaredLimits);
const specDeclaredLimits = new Set(matches(/^- \*\*(S-[A-Z0-9-]+)\*\*.*Declared limit .*NOT a test case/gm, spec));
const sameSet = (left, right) => left.size === right.size && [...left].every(value => right.has(value));
if (!sameSet(declaredLimits, specDeclaredLimits)) fail(`declared-limit manifest differs from SPEC: manifest=${JSON.stringify([...declaredLimits].sort())}, spec=${JSON.stringify([...specDeclaredLimits].sort())}`);
const unknownLimits = [...declaredLimits].filter(value => !cases.has(value)).sort();
if (unknownLimits.length > 0) fail(`declared limits absent from SPEC: ${unknownLimits.join(", ")}`);

const assignments = new Map([...cases].map(value => [value, []]));
const emptyPatterns = [];
const phaseCounts = new Map();
for (const [phase, rawPatterns] of Object.entries(config.phases)) {
  const patterns = rawPatterns.map(raw => ({ raw, regex: new RegExp(`^(?:${raw})$`) }));
  for (const { raw, regex } of patterns) if (![...cases].some(value => regex.test(value))) emptyPatterns.push(`${phase}:${raw}`);
  for (const value of cases) if (patterns.some(({ regex }) => regex.test(value))) assignments.get(value).push(phase);
  phaseCounts.set(phase, [...assignments.values()].filter(owners => owners.includes(phase)).length);
}
if (emptyPatterns.length > 0) fail(`patterns matching no case: ${emptyPatterns.join(", ")}`);
const missing = [...assignments].filter(([, owners]) => owners.length === 0).map(([value]) => value).sort();
const multiple = [...assignments].filter(([, owners]) => owners.length > 1).sort(([left], [right]) => left.localeCompare(right));
if (missing.length > 0) fail(`unassigned cases: ${missing.join(", ")}`);
if (multiple.length > 0) fail(`multiply assigned cases: ${multiple.map(([value, owners]) => `${value}=${JSON.stringify(owners)}`).join(", ")}`);

const testCount = cases.size - declaredLimits.size;
if (testCount !== config.expectedTestCases) fail(`expected ${String(config.expectedTestCases)} test cases, found ${String(testCount)}`);
const counts = [...phaseCounts].map(([phase, count]) => `${phase}=${String(count)}`).join(" ");
process.stdout.write(`implementation phases OK: ${String(cases.size)} definitions, ${String(testCount)} tests, ${String(declaredLimits.size)} declared limit; ${counts}\n`);
